### DISCUSSION TITLE
Allow change validation for "Reference" field

### DISCUSSION THREAD

**Discussion by abramofranchetti** [PrestaShop/PrestaShop#40141] [Ideas]
### Prerequisites

- [X] I understand and accept the project's [code of conduct](https://github.com/PrestaShop/PrestaShop/blob/develop/CODE_OF_CONDUCT.md)
- [X] I have already [searched in existing features request](https://github.com/PrestaShop/PrestaShop/issues?q=is%3Aissue+label%3AImprovement%2CFeature%2CDeveloper+Feature+) and found no previous suggestion of this feature.

### Is your feature request related to a problem?

I'm here to talk about the 'reference' field of products and combinations.
At the moment this field has as legacy validation "isReference" and in the new core a new type "Reference" has been made that takes care of the validation.

The same validation is used for all types of 'references':
that of the product
that of the combination
that of the order
that of suppliers
etc. etc

At least as regards the product (but also for the other fields, except perhaps the order reference) the validation is too stringent, the current pattern for the regex is '/^[^<>;={}]*$/u'
I have worked with ERP and management software for the last 15 years and I assure you that the management code of an article can contain the most disparate characters. ES: =,!,> <; . # etc. etc.

Now in the new symfony system it has become impossible to remove the validation of the 'reference' field for the product and combination class because it has been entrusted to the core class "Reference" and therefore requires a modification to the core.

### Describe the solution you'd like

Make the validation less severe and distinguish between the validation of the reference of the various entities.
Furthermore allow to modify the validation of 'PrestaShop\PrestaShop\Core\Domain\Product\ValueObject.Reference' (a new hook?) because now i can override Validate::isReference but cannot modify the validation or the Reference Class. 

Also i dont understand because the Order::reference have to be validated with the same logic of a Product::reference that is something that come often from ERP software...
Idem for the Stock::reference of SupplyOrderDetail::reference.

But first of all would be necessary to allow change of validation without touch the core class Reference.

### Alternatives you've considered

Remove validation for reference in product or make it such as 'isString'. Reference is often a string that come from ERP and other management software so can be everithin and often this cannot be changed because company rules ecc.

OR

make an hook in the Reference class and Validate::isReference method that allow to change the pattern string for the regex validation.

### Additional context

example of product reference that i find every other day working with ERP interfacing.

PK-ART=1.5/1
1%2.1547
.1445890999
****123456
16;18;1987#2

### Do you plan to work on this feature?

- [X] I'm willing to contribute a [formal specification](https://github.com/PrestaShop/prestashop-specs).
- [X] I'm willing to provide any wireframes or design assets required for this feature.
- [X] I'm willing to submit a Pull Request that implements this feature.
- [X] I'm willing to help verify that the implemented feature works as intended and produces no unintended side effects.

**Comment by sLorenzini** [comment]
Hello @abramofranchetti,

Thank you for your suggestion. The Product Team will take it into consideration for future developments. 

Please be aware that there is no guarantee that this feature will be developed anytime soon. 

We’re waiting for your PR 🚀

Thank you
